File: android/src/main/java/com/mytracker/MytrackerModule.java
package com.mytracker;

import android.app.Application;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.my.tracker.MyTracker;
import com.my.tracker.MyTrackerConfig;
import com.my.tracker.MyTrackerParams;

import java.util.HashMap;
import java.util.Map;

@ReactModule(name = MytrackerModule.NAME)
public class MytrackerModule extends ReactContextBaseJavaModule {
  public static final String NAME = "MyTracker";
  private final Application application;

  public MytrackerModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.application = (Application) reactContext.getApplicationContext();
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  @ReactMethod
  public void initTracker(String SDK_KEY) {
    MyTracker.initTracker(SDK_KEY, application);
  }

  @ReactMethod
  public void trackEvent(String name) {
    MyTracker.trackEvent(name);
  }

  @ReactMethod
  public void trackLoginEvent(String userId, String vkConnectId) {
    MyTracker.trackLoginEvent(userId, vkConnectId);
  }

  @ReactMethod
  public void trackLoginEventWithParams(String userId, String vkConnectId, ReadableMap attributes) {
    Map<String, Object> map = attributes.toHashMap();
    Map<String, String> params = new HashMap<>();
    for (Map.Entry<String, Object> entry : map.entrySet()) {
      if (entry.getValue() instanceof String) {
        params.put(entry.getKey(), (String) entry.getValue());
      }
    }
    MyTracker.trackLoginEvent(userId, vkConnectId, params);
  }

  @ReactMethod
  public void trackEventWithParams(String name, ReadableMap attributes) {
    Map<String, Object> map = attributes.toHashMap();
    Map<String, String> params = new HashMap<>();
    for (Map.Entry<String, Object> entry : map.entrySet()) {
      if (entry.getValue() instanceof String) {
        params.put(entry.getKey(), (String) entry.getValue());
      }
    }
    MyTracker.trackEvent(name, params);
  }

  @ReactMethod
  public void setCustomUserId(String USER_ID) {
    MyTrackerParams trackerParams = MyTracker.getTrackerParams();
    trackerParams.setCustomUserId(USER_ID);
  }

  @ReactMethod
  public void trackInviteEvent() {
    MyTracker.trackInviteEvent();
  }

  @ReactMethod
  public void trackInviteEventWithParams(ReadableMap attributes) {
    Map<String, Object> map = attributes.toHashMap();
    Map<String, String> params = new HashMap<>();
    for (Map.Entry<String, Object> entry : map.entrySet()) {
      if (entry.getValue() instanceof String) {
        params.put(entry.getKey(), (String) entry.getValue());
      }
      MyTracker.trackInviteEvent(params);
    }
  }

  @ReactMethod
  public void flush() {
    MyTracker.flush();
  }

  @ReactMethod
  public void trackRegistrationEvent(String userId, String vkConnectId) {
    MyTracker.trackRegistrationEvent(userId, vkConnectId);
  }

  @ReactMethod
  public void trackRegistrationEventWithParams(String userId, String vkConnectId, ReadableMap attributes) {
    Map<String, Object> map = attributes.toHashMap();
    Map<String, String> params = new HashMap<>();
    for (Map.Entry<String, Object> entry : map.entrySet()) {
      if (entry.getValue() instanceof String) {
        params.put(entry.getKey(), (String) entry.getValue());
      }
    }
    MyTracker.trackRegistrationEvent(userId, vkConnectId, params);
  }

  @ReactMethod
  public void trackLevel() {
    MyTracker.trackLevelEvent();
  }

  @ReactMethod
  public void trackLevelWithLevel(int level) {
    MyTracker.trackLevelEvent(level, (Map<String, String>) null);
  }

  @ReactMethod
  public void trackLevelWithLevelWithParams(int level, ReadableMap attributes) {
    Map<String, Object> map = attributes.toHashMap();
    Map<String, String> params = new HashMap<>();
    for (Map.Entry<String, Object> entry : map.entrySet()) {
      if (entry.getValue() instanceof String) {
        params.put(entry.getKey(), (String) entry.getValue());
      }
    }
    MyTracker.trackLevelEvent(level, params);
  }

  @ReactMethod
  public void trackLaunchEnable(boolean enable) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    trackerConfig.setTrackingLaunchEnabled(enable);
  }

  @ReactMethod
  public void trackLaunchTimeout(int seconds) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    trackerConfig.setLaunchTimeout(seconds);
  }

  @ReactMethod
  public void bufferingPeriod(int seconds) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    trackerConfig.setBufferingPeriod(seconds);
  }

  @ReactMethod
  public void forcingPeriod(int seconds) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    trackerConfig.setForcingPeriod(seconds);
  }

  @ReactMethod
  public void autotrackPurchase(boolean enable) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    trackerConfig.setAutotrackingPurchaseEnabled(enable);
  }

  @ReactMethod
  public void trackLocation(int number) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    if (number == 0) {
      trackerConfig.setTrackingLocationEnabled(false);
    }
    if (number == 1 || number == 2) {
      trackerConfig.setTrackingLocationEnabled(true);
    }
  }

  @ReactMethod
  public void region(int number) {
    MyTrackerConfig trackerConfig = MyTracker.getTrackerConfig();
    if (number == 1) {
      trackerConfig.setRegion(MyTrackerConfig.Region.RU);
    }
    if (number == 2) {
      trackerConfig.setRegion(MyTrackerConfig.Region.EU);
    }
  }

  @ReactMethod
  public void setDebugMode(boolean enable) {
    MyTracker.setDebugMode(enable);
  }
}

File: android/src/main/java/com/mytracker/MytrackerModule.kt
package com.mytracker

import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.bridge.ReactApplicationContext
import android.app.Application
import com.my.tracker.MyTracker
import com.facebook.react.bridge.ReadableMap
import java.util.HashMap

@ReactModule(name = MytrackerModule.NAME)
class MytrackerModule(reactContext: ReactApplicationContext) :
    NativeMytrackerSpec(reactContext) {
    private val application: Application

    init {
        application = reactContext.applicationContext as Application
    }

    override fun getName(): String {
        return NAME
    }

    override fun initTracker(SDK_KEY: String) {
        MyTracker.initTracker(SDK_KEY, application)
    }


   override fun trackEvent(name: String) {
        MyTracker.trackEvent(name)
    }


   override fun trackLoginEvent(userId: String, vkConnectId: String?) {
        MyTracker.trackLoginEvent(userId, vkConnectId)
    }


  override fun trackLoginEventWithParams(userId: String, vkConnectId: String?, attributes: ReadableMap) {
        val map: HashMap<String, Any?> = attributes.toHashMap()
        val params: MutableMap<String, String> = HashMap()
        for ((key, value) in map) {
            if (value is String) {
                params[key] = value
            }
        }
        MyTracker.trackLoginEvent(userId, vkConnectId, params)
    }


   override fun trackEventWithParams(name: String, attributes: ReadableMap) {
        val map: HashMap<String, Any?> = attributes.toHashMap()
        val params: MutableMap<String, String> = HashMap()
        for ((key, value) in map) {
            if (value is String) {
                params[key] = value
            }
        }
        MyTracker.trackEvent(name, params)
    }


   override fun setCustomUserId(USER_ID: String?) {
        val trackerParams = MyTracker.getTrackerParams()
        trackerParams.setCustomUserId(USER_ID)
    }


   override fun trackInviteEvent() {
        MyTracker.trackInviteEvent()
    }


   override fun trackInviteEventWithParams(attributes: ReadableMap) {
        val map: HashMap<String, Any?> = attributes.toHashMap()
        val params: MutableMap<String, String> = HashMap()
        for ((key, value) in map) {
            if (value is String) {
                params[key] = value
            }
            MyTracker.trackInviteEvent(params)
        }
    }


   override fun flush() {
        MyTracker.flush()
    }


   override fun trackRegistrationEvent(userId: String, vkConnectId: String?) {
        MyTracker.trackRegistrationEvent(userId, vkConnectId)
    }


   override fun trackRegistrationEventWithParams(
        userId: String,
        vkConnectId: String?,
        attributes: ReadableMap
    ) {
        val map: HashMap<String, Any?> = attributes.toHashMap()
        val params: MutableMap<String, String> = HashMap()
        for ((key, value) in map) {
            if (value is String) {
                params[key] = value
            }
        }
        MyTracker.trackRegistrationEvent(userId, vkConnectId, params)
    }


   override fun trackLevel() {
        MyTracker.trackLevelEvent()
    }


   override fun trackLevelWithLevel(level: Double) {
        MyTracker.trackLevelEvent(level.toInt(), null as Map<String?, String?>?)
    }


   override fun trackLevelWithLevelWithParams(level: Double, attributes: ReadableMap) {
        val map: HashMap<String, Any?> = attributes.toHashMap()
        val params: MutableMap<String, String> = HashMap()
        for ((key, value) in map) {
            if (value is String) {
                params[key] = value
            }
        }
        MyTracker.trackLevelEvent(level.toInt(), params)
    }


   override fun trackLaunchEnable(enable: Boolean) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.isTrackingLaunchEnabled = enable
    }


   override fun trackLaunchTimeout(seconds: Double) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.launchTimeout = seconds.toInt()
    }


   override fun bufferingPeriod(seconds: Double) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.bufferingPeriod = seconds.toInt()
    }


   override fun forcingPeriod(seconds: Double) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.forcingPeriod = seconds.toInt()
    }


   override fun autotrackPurchase(enable: Boolean) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.isAutotrackingPurchaseEnabled = enable
    }


   override fun trackLocation(number: Double) {
        val trackerConfig = MyTracker.getTrackerConfig()
        trackerConfig.setLocationTrackingMode(number.toInt())
    }


   override fun setDebugMode(enable: Boolean) {
        MyTracker.setDebugMode(enable)
    }

   override fun getInstanceId(): String {
      val id = MyTracker.getInstanceId(reactApplicationContext)
      return id
    }

  companion object {
        const val NAME = "Mytracker"
    }
}
